crates/node/builder/src/rpc.rs

```diff
@@ -292,102 +296,232 @@ where
-    EthApi: EthApiBuilderProvider<Node> + FullEthApiServer,
+    RpcRegistry<Node, EthApi>: Debug,
 {
-    let auth_config = config.rpc.auth_server_config(jwt_secret)?;
-    let module_config = config.rpc.transport_rpc_module_config();
-    debug!(target: "reth::cli", http=?module_config.http(), ws=?module_config.ws(), "Using RPC module config");
-
-    let (mut modules, mut auth_module, registry) = RpcModuleBuilder::default()
-        .with_provider(node.provider().clone())
-        .with_pool(node.pool().clone())
-        .with_network(node.network().clone())
-        .with_events(node.provider().clone())
-        .with_executor(node.task_executor().clone())
-        .with_evm_config(node.evm_config().clone())
-        .with_block_executor(node.block_executor().clone())
-        .build_with_auth_server(module_config, engine_api, EthApi::eth_api_builder());
-
-    let mut registry = RpcRegistry { registry };
-    let ctx = RpcContext {
-        node: node.clone(),
-        config,
-        registry: &mut registry,
-        modules: &mut modules,
-        auth_module: &mut auth_module,
-    };
-
-    let RpcAddOns { hooks, .. } = add_ons;
-    let RpcHooks { on_rpc_started, extend_rpc_modules } = hooks;
-
-    extend_rpc_modules.extend_rpc_modules(ctx)?;
-
-    let server_config = config.rpc.rpc_server_config();
-    let cloned_modules = modules.clone();
-    let launch_rpc = server_config.start(&cloned_modules).map_ok(|handle| {
-        if let Some(path) = handle.ipc_endpoint() {
-            info!(target: "reth::cli", %path, "RPC IPC server started");
-        }
-        if let Some(addr) = handle.http_local_addr() {
-            info!(target: "reth::cli", url=%addr, "RPC HTTP server started");
-        }
-        if let Some(addr) = handle.ws_local_addr() {
-            info!(target: "reth::cli", url=%addr, "RPC WS server started");
-        }
-        handle
-    });
-
-    let launch_auth = auth_module.clone().start_server(auth_config).map_ok(|handle| {
-        let addr = handle.local_addr();
-        if let Some(ipc_endpoint) = handle.ipc_endpoint() {
-            info!(target: "reth::cli", url=%addr, ipc_endpoint=%ipc_endpoint,"RPC auth server started");
-        } else {
-            info!(target: "reth::cli", url=%addr, "RPC auth server started");
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RpcHandle")
+            .field("rpc_server_handles", &self.rpc_server_handles)
+            .field("rpc_registry", &self.rpc_registry)
+            .finish()
+    }
+}
+
+/// Node add-ons containing RPC server configuration, with customizable eth API handler.
+#[allow(clippy::type_complexity)]
+pub struct RpcAddOns<Node: FullNodeComponents, EthApi: EthApiTypes> {
+    /// Additional RPC add-ons.
+    pub hooks: RpcHooks<Node, EthApi>,
+    /// Builder for `EthApi`
+    eth_api_builder: Box<dyn FnOnce(&EthApiBuilderCtx<Node>) -> EthApi + Send + Sync>,
+    _pd: PhantomData<(Node, EthApi)>,
+}
```

Comment:
cool, with this we're also unblocked to convert the eth_api_builder into something nicer